architecture/2025-04-compute-extensibility.md

```diff
@@ -0,0 +1,474 @@
+* **Clear architectural separation**: This plan implements a clear architectural separation between core functionality (environments and applications) and platform-specific provisioning (UDTs and recipes).
+
+### Disadvantages vs Alternate Plans
+
+* **Changed application model**: This plan results in a more consistent application model, but it is a change from the existing application model, which may cause confusion and upgrade work for users.
+* **Higher initial complexity**: Implementing core types as UDTs requires solving more problems before having a releasable feature.
+* **Delayed delivery**: Initial capabilities will take longer to deliver a release.
+
+## Alternate Plan: Add Recipe Support to Existing Core Types
+
+The alternate plan is to keep the core types (`containers`, `gateways`, and `secretStores`) as built-in types in the Radius application model, and add recipe support to them.
+
+| Phase | Name | Size | Activities | Customer Capabilities |
+| ----- | ---- | ---- | ---------- | --------------------- |
+| 1 | Support Recipes on Core Types | M | Enable recipes on new versions of core types | Recipes can be registered for core resource types. |
+| | Release | | | |
+| 2 | ACI recipes | M | Implement ACI provisioning from recipes |- ACI is deployed via default recipes<br> - ACI recipes can be modified/replaced by customers |
+| | Release | | | |
+| 3 | Kubernetes recipes | L | - Implement Kubernetes provisioning from recipes<br>- Remove existing Kubernetes provisioning code |- Kubernetes is deployed via default recipes<br> - Kubernetes recipes can be modified/replaced by customers |
+| | Release | | | |
+
+### Advantages vs Recommended Plan
+
+* **Stable application model**: Maintains the familiar Radius application model that customers are already using.
+* **Incremental change**: Adding recipe support to core types is an additive change, and adding ACI provisioning with recipes is not disruptive to the existing Kubernetes provisioning code.
+* **Earlier delivery**: ACI recipe provisioning can be added and released before Kubernetes recipe is released.
+
+### Disadvantages vs Recommended Plan
+
+* **Less architectural flexibility**: Core types remain hard-coded in Radius, limiting some extensibility options, e.g., the ability of a customer to copy and modify a core type.
+* **Inconsistent resource type model**: Core types and UDTs would have different implementation approaches. However, customers can choose to ignore the core types and implement their own.
+
+## Open Questions
+
+* What is the impact on the Radius graph, and how would we continue to support it? We will could implement the graph as the relationships defined in Bicep, or implement connections in a Bicep/Terraform reusable module.
+* Is the Radius group concept affected by this design?
+* Can everything currently deployed by Go code be deployed using recipes? We think so, but need to prove it through prototyping.
```

Comment:
How will this (moving the logic from Go code into Bicep/Terraform) affect the performance? Should we think about this too?

Reply:
I don't think there will be a large effect because most of the time spent during provisioning is waiting for resources to be created on the target platform. But we should measure it.